### PR TITLE
added URL parser

### PR DESCRIPTION
**Pull request by yuvalherziger**
see https://github.com/mpscholten/request-parser/issues/17

### CODE REVIEW

**Comment by yuvalherziger** [comment]
@mpscholten , I was wondering about the design. We could perhaps do something like this, instead of an additional parser (which extends `StringParser` at the moment):

`StringParser` could look like so:

``` php
class StringParser extends AbstractValueParser
{
    private $validation;

    public function __construct(callable $exceptionFactory, $name, $value, $validation = null)
    {
        parent::__construct($exceptionFactory, $name, $value);
        $this->validation = $validation;
    }

    protected function parse($value)
    {
        if ($this->validation === FILTER_VALIDATE_URL && filter_var($value, FILTER_VALIDATE_URL) === false) {
            return null;
        }
        return (string) $value;
    }
...
```

Namely, `TypeParser` could have the following method:

``` php
public function url()
{
    return new StringParser($this->exceptionFactory, $this->name, $this->value, FILTER_VALIDATE_URL);
}
```

Which approach do you like better?
Same thing applies for https://github.com/mpscholten/request-parser/pull/21 😃

**Comment by mpscholten** [comment]
I'd rather pick inheritance because then we can show better error messages like "Invalid email address". But to do this we first have to refactor the exception hierarchy :) We should have atleast another `InvalidInputException` (which has to extend the `ResourceNotFound` root exception for b.c. reasons)

**Comment by mpscholten** [comment]
What do you think about making the API to look like `$this->bodyParameter('url')->string()->url()->required()`? The set of all emails is a subset of all strings. Also it shows that the return value is not some `URL` object but rather a string.

**Comment by yuvalherziger** [comment]
Good point.

> Also it shows that the return value is not some URL object but rather a string

We could perhaps parse a URL fully, and then you get an array with the components
http://php.net/manual/en/function.parse-url.php
Or alternatively use a 3rd party, e.g. http://url.thephpleague.com/overview/

**Comment by mpscholten** [comment]
While it's useful, it would add a new dependency to the library and I think that would be overkill because parsing urls is not used so often (atleast compared to strings, ints and the other parsers)

**Comment by yuvalherziger** [comment]
Yes. The only question is, now, whether or not `[]` is too general of a return type.

**Comment by mpscholten** [comment]
Ah okay, you would return the result of `parse_url`? Actually I wouldn't parse the url, as most likely that's not what the user wants. E.g. an example use case could be something like:

``` php
<?php
function redirectAction() {
    $targetUrl = $this->queryParameter('to')->string()->url()->required();
    return new RedirectResponse($targetUrl);
}
```

If the user wants e.g. only the hostname, he can call `parse_url` himself (or use the library of his choice).

**Comment by yuvalherziger** [comment]
We can somehow do this, but in a way that we prevent `$this->queryParameter('to')->string()->url()->url()->url()->url()`, so `url()` cant really by a member method of `StringParser`

**Comment by mpscholten** [comment]
Yup, `url` should return `new UrlParser($this->name, $this->value)`. This way you cannot chain the `url` calls.

**Comment by yuvalherziger** [comment]
Oh nice. OK, so the only things I'll also need to amend are AbstractValueParser's `exceptionFactory` and `name` from private to protected (in order to use them for an instance of a new parser from another parser (i.e. `StringParse` ---> `UrlParser`). And also to fully prevent chaining, UrlParser won't extend StringParser :) I'm on it

**Comment by mpscholten** [comment]
Awesome thanks :)

**Comment by yuvalherziger** [comment]
ready for review 👍

File: examples/psr7.php [hunk 1]
```diff
@@ -50,6 +51,13 @@ public function json()
 
         return print_r($payload, true);
     }
+
+    public function url()
```
Comment (mpscholten) [inline examples/psr7.php]: Maybe we can keep out the examples :) They're intended to show just the basic operations

Comment (yuvalherziger) [inline examples/psr7.php]: done :)

**Pull request merged**